game/src/OptionMenu/ResolutionSelector.gd

```diff
@@ -63,7 +71,47 @@ func _set_value_from_file(load_value):
 
 func _on_item_selected(index : int):
 	if _valid_index(index):
+		previous_resolution = Resolution.get_current_resolution()
 		Resolution.set_resolution(get_item_metadata(index))
+		var new_resolution = get_item_metadata(index)
+		#emit a signal if this option was changed by the user
+		if was_pressed and index != previous_selected_index:
+			emit_signal("selection_changed",previous_resolution,new_resolution)
 	else:
 		push_error("Invalid ResolutionSelector index: %d" % index)
 		reset_setting()
+
+#Dialog revert functionality below here
+#UIFUN-301 and UIFUN-302
+
+#This is just to filter out cases where _on_item_selected is not called by the
+#user changing a setting (which happens on startup)
+var was_pressed:bool = false
+func _on_toggled(button_pressed):
+	was_pressed = !button_pressed
+	
+func _on_selection_changed(previous_resolution, _new_resolution):
+	if revert_dialog:
+		print("Start Revert Countdown!")
+		start_revert_countdown()
+		
+func _process(_delta):
+	var time_left:int = -1
+	if not timer.is_stopped():
+		time_left = timer.time_left
+	revert_dialog.dialog_text = tr("OPTIONS_VIDEO_RESOLUTION_DIALOG_TEXT").format({"time":time_left})
+
+func start_revert_countdown() -> void:
+	timer.start(countdown_seconds)
+	revert_dialog.popup_centered(Vector2(1,1))
+
+func _on_confirmed() -> void:
+	timer.stop()
+	revert_dialog.hide()
+
+func _cancel_changes() -> void:
+	Resolution.set_resolution(previous_resolution)
+	_sync_resolutions()
+	print("Resolution reset to (%dx%d)" % [previous_resolution.x,previous_resolution.y])
+	timer.stop()
+	revert_dialog.hide()
```

Comment:
This is unnecessary when a dialog's dialog_hide_on_ok toggle is on. (which is default for most dialogs)

Reply:
From experimentation, this seems necessary since this function is triggered both when the timer stops (which means the dialog needs to be hidden manually), and the dialog's cancel option (which requires the timer be stopped.

Reply:
Cancel should be hiding the dialog as well.

Reply:
One way this could work would be if you could emit the `pressed` signal on `ConfirmationDialog.get_cancel_button()` for the Timer's timeout, you can't do this in the scene but you could set this in the `_ready` function. If there are more places where this pops up, I think I'll come back to write a `TimedConfirmationDialog` node to handles this instead, but for now it shouldn't matter.